scala-xml/src/main/scala/scalaxml/ElemInstances.scala

```diff
@@ -17,22 +18,21 @@ trait ElemInstances {
   /**
    * Handles a message body as XML.
    *
-   * TODO Not an ideal implementation.  Would be much better with an asynchronous XML parser, such as Aalto.
-   *
    * @param parser the SAX parser to use to parse the XML
    * @return an XML element
    */
   implicit def xml(implicit parser: SAXParser = XML.parser): EntityDecoder[Elem] = {
     import EntityDecoder._
-    decodeBy(MediaType.`text/xml`, MediaType.`text/html`, MediaType.`application/xml`){ msg =>
-      collectBinary(msg).flatMap[Elem] { arr =>
-        val source = new InputSource(new StringReader(new String(arr.toArray, msg.charset.getOrElse(Charset.`US-ASCII`).nioCharset)))
-        try DecodeResult.success(Task.now(XML.loadXML(source, parser)))
-        catch {
-          case e: SAXParseException =>
-            DecodeResult.failure(MalformedMessageBodyFailure("Invalid XML", Some(e)))
-          case NonFatal(e) => DecodeResult(Task.fail(e))
-        }
+    decodeBy(MediaType.`text/xml`, MediaType.`text/html`, MediaType.`application/xml`) { msg =>
+      val stream = toInputStream(msg.body)
+      val source = new InputSource(stream)
+      msg.charset.foreach(cs => source.setEncoding(cs.nioCharset.name))
```

Comment:
I believe the reason is that the default charset for XML is US-ASCII.
https://www.ietf.org/rfc/rfc3023.txt